pre_commit/commands/try_repo.py

```diff
@@ -12,23 +13,50 @@
 from pre_commit.clientlib import load_manifest
 from pre_commit.commands.run import run
 from pre_commit.store import Store
+from pre_commit.util import cmd_output
 from pre_commit.util import tmpdir
 
+logger = logging.getLogger(__name__)
 
-def try_repo(args):
-    ref = args.ref or git.head_rev(args.repo)
 
+def _repo_ref(tmpdir, repo, ref):
+    # if `ref` is explicitly passed, use it
+    if ref:
+        return repo, ref
+
+    ref = git.head_rev(repo)
+    # if it exists on disk, we'll try and clone it with the local changes
+    if os.path.exists(repo) and git.has_diff('HEAD', repo=repo):
+        logger.warning('Creating temporary repo with uncommitted changes...')
+
+        shadow = os.path.join(tmpdir, 'shadow-repo')
+        cmd_output('git', 'clone', '--no-checkout', repo, shadow)
+        cmd_output('git', 'checkout', ref, '-b', '_pc_tmp', cwd=shadow)
+        idx = git.git_path('index', repo=shadow)
+        objs = git.git_path('objects', repo=shadow)
+        env = dict(os.environ, GIT_INDEX_FILE=idx, GIT_OBJECT_DIRECTORY=objs)
+        cmd_output('git', 'add', '-u', cwd=repo, env=env)
```

Comment:
I'm most concerned about these ~4 lines.  this mucks with git internals to make a commit into a shadow repo.  it seems to work, but I can't really tell if this is intended to work or not